openedx/openedx-platform
# Define custom completion for VideoModule

Update the VideoModule to publish a completion event when the player
reaches 95% complete, and submit a BlockCompletion when that event
occurs.

## Testing instructions

1.  In the lms shell, remove all existing BlockCompletions for video blocks.
    ```python
    from lms.djangoapps.completion.models import BlockCompletion
    BlockCompletion.objects.filter(block_type='video').delete()
    ```
2.  Sign in to the LMS and visit the demo course.
3.  Navigate to "Example Week 1: Getting Started > Lesson 1 - Getting Started > (Vertical 2) Working with Videos".  Open the browser's development panel, and view network traffic.  
4. Verify that when the video reach 4m 3s, a POST is sent to a URL for the block ending with "publish_completion", and receives a 200 response.  
5.  Verify that a BlockCompletion now exists for the video block with completion value == 1.0.
    ```python
    from lms.djangoapps.completion.models import BlockCompletion
    print(BlockCompletion.objects.filter(block_type='video', completion=1.0))
    ```
6.  Navigate to the next vertical: "Videos on edX".  This one uses more features (a custom start time and end-time).  Watch to the end of the video (or jump to a few seconds before the end of the video, and just watch the end). 
7.  Verify that a publish_completion POST is sent after 7m 17s (Start time is 5m 10s, end time is 7m 24s = 134s; 134 * .95 = 127.3s; 5m 10s + 127s = 7m 17s), and receives a 200 response.
8.  Verify that a BlockCompletion object exists for this block as well.

## Author concerns

I'm rusty at Javascript, and entirely new to the video player.  I may not be doing things the right way. 

## Reviewers
- [x] @robrap 

FYI: @yro for video player expertise.

## TODO:

- [x] Rebase PR against edx:master after cliff/complete-default-blocks merges ( #16234 )
